sessionctx/sessionstates/session_token.go

```diff
@@ -0,0 +1,338 @@
+	// certs contain all the certificates that are not expired yet.
+	certs []*certInfo
+}
+
+type certInfo struct {
+	cert       *x509.Certificate
+	privKey    crypto.PrivateKey
+	expireTime time.Time
+}
+
+// We cannot guarantee that the cert and key paths are set at the same time because they are set through system variables.
+func (sc *signingCert) setCertPath(certPath string) {
+	sc.Lock()
+	// Just in case of repeatedly loading global variables, we check the path to avoid useless loading.
+	if certPath != sc.certPath {
+		sc.certPath = certPath
+		// It may fail expectedly because the key path is not set yet.
+		sc.checkAndLoadCert()
+	}
+	sc.Unlock()
+}
+
+func (sc *signingCert) setKeyPath(keyPath string) {
+	sc.Lock()
+	if keyPath != sc.keyPath {
+		sc.keyPath = keyPath
+		// It may fail expectedly because the cert path is not set yet.
+		sc.checkAndLoadCert()
+	}
+	sc.Unlock()
+}
+
+func (sc *signingCert) lockAndLoad() {
+	sc.Lock()
+	sc.checkAndLoadCert()
+	sc.Unlock()
+}
+
+func (sc *signingCert) checkAndLoadCert() {
+	if len(sc.certPath) == 0 || len(sc.keyPath) == 0 {
+		return
+	}
+	if err := sc.loadCert(); err != nil {
+		logutil.BgLogger().Warn("loading signing cert failed",
+			zap.String("cert path", sc.certPath),
+			zap.String("key path", sc.keyPath),
+			zap.Error(err))
+	} else {
+		logutil.BgLogger().Info("signing cert is loaded successfully",
+			zap.String("cert path", sc.certPath),
+			zap.String("key path", sc.keyPath))
+	}
+}
+
+// loadCert loads the cert and adds it into the cert list.
+func (sc *signingCert) loadCert() error {
+	tlsCert, err := tls.LoadX509KeyPair(sc.certPath, sc.keyPath)
+	if err != nil {
+		return errors.Wrapf(err, "load x509 failed, cert path: %s, key path: %s", sc.certPath, sc.keyPath)
+	}
+	var cert *x509.Certificate
+	if tlsCert.Leaf != nil {
+		cert = tlsCert.Leaf
+	} else {
+		if cert, err = x509.ParseCertificate(tlsCert.Certificate[0]); err != nil {
+			return errors.Wrapf(err, "parse x509 cert failed, cert path: %s, key path: %s", sc.certPath, sc.keyPath)
+		}
+	}
+
+	// rotate certs
+	now := getNow()
+	newCerts := make([]*certInfo, 0, len(sc.certs)+1)
+	newCerts = append(newCerts, &certInfo{
+		cert:       cert,
+		privKey:    tlsCert.PrivateKey,
+		expireTime: now.Add(loadCertInterval + oldCertValidTime),
+	})
+	for i := 0; i < len(sc.certs); i++ {
+		if now.After(sc.certs[i].expireTime) {
+			break
+		}
+		newCerts = append(newCerts, sc.certs[i])
+	}
+	sc.certs = newCerts
+	return nil
+}
+
+// sign generates a signature with the content and the private key.
+func (sc *signingCert) sign(content []byte) ([]byte, error) {
+	var (
+		signer crypto.Signer
+		opts   crypto.SignerOpts
+	)
+	sc.RLock()
+	defer sc.RUnlock()
+	if len(sc.certs) == 0 {
+		return nil, errors.New("no certificate or key file to sign the data")
+	}
+	// Always sign the token with the latest cert.
+	certInfo := sc.certs[0]
+	switch certInfo.privKey.(type) {
```

Comment:
I think we could do `t := xxx.(type)` to eliminate the manual type casting ` certInfo.privKey.(ed25519.PrivateKey)`.

Reply:
PTAL again, is this expected now?